pkg/validator/validators.go

```diff
@@ -244,25 +254,149 @@ func (kc *policyValidator) upstreamTrafficSettingValidator(req *admissionv1.Admi
 		}
 	}
 	for _, route := range upstreamTrafficSetting.Spec.HTTPRoutes {
 		if route.RateLimit != nil && route.RateLimit.Local != nil {
 			if _, ok := xds_type.StatusCode_name[int32(route.RateLimit.Local.ResponseStatusCode)]; !ok {
-				return nil, fmt.Errorf("Invalid responseStatusCode %d. See https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/v3/http_status.proto#enum-type-v3-statuscode for allowed values",
+				return nil, fmt.Errorf("invalid responseStatusCode %d. See https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/v3/http_status.proto#enum-type-v3-statuscode for allowed values",
 					route.RateLimit.Local.ResponseStatusCode)
 			}
 		}
 	}
 
 	return nil, nil
 }
+
+var validMRCTransitions = map[string][]string{
+	constants.MRCStateValidatingRollout:  {constants.MRCStateIssuingRollout, constants.MRCStateError},
+	constants.MRCStateIssuingRollout:     {constants.MRCStateActive, constants.MRCStateError},
+	constants.MRCStateActive:             {constants.MRCStateValidatingRollback, constants.MRCStateError},
+	constants.MRCStateValidatingRollback: {constants.MRCStateIssuingRollback, constants.MRCStateError},
+	constants.MRCStateIssuingRollback:    {constants.MRCStateInactive, constants.MRCStateError},
+	constants.MRCStateInactive:           {},
+	constants.MRCStateError:              {},
+	"":                                   {constants.MRCStateValidatingRollout, constants.MRCStateError},
+}
+
+// meshRootCertificateValidator validates the MeshRootCertificate custom resource
+func (cv *configValidator) meshRootCertificateValidator(req *admissionv1.AdmissionRequest) (*admissionv1.AdmissionResponse, error) {
+	log.Debug().Msg("made it to MRC validator")
+	switch req.Operation {
+	case admissionv1.Create:
+		newMRC := &configv1alpha2.MeshRootCertificate{}
+		if err := json.NewDecoder(bytes.NewBuffer(req.Object.Raw)).Decode(newMRC); err != nil {
+			return nil, err
+		}
+
+		if newMRC.Spec.TrustDomain == "" {
+			return nil, fmt.Errorf("trustDomain must be non empty for MRC %s/%s", newMRC.GetNamespace(), newMRC.GetName())
+		}
+
+		// if no state is set, MRC will not be considered for rotation
+		if newMRC.Status.State == "" {
+			return nil, nil
+		}
+
+		if cv.activeRootCertificateRotation(newMRC) {
+			return nil, fmt.Errorf("cannot create MRC %s/%s with status %s. Certificate rotation in progress. At most, only 2 MRCs can be in non error or non inactive states",
+				newMRC.GetNamespace(), newMRC.GetName(), newMRC.Status.State)
+		}
+
+		if _, ok := validMRCTransitions[newMRC.Status.State]; !ok {
+			return nil, fmt.Errorf("invalid status %s for MRC %s/%s. Must be %s, %s, %s, %s, %s, %s, or %s",
+				newMRC.Status.State, newMRC.GetNamespace(), newMRC.GetName(), constants.MRCStateValidatingRollout,
+				constants.MRCStateIssuingRollout, constants.MRCStateActive, constants.MRCStateIssuingRollback,
+				constants.MRCStateValidatingRollback, constants.MRCStateInactive, constants.MRCStateError)
+		}
+	case admissionv1.Update:
```

Comment:
A thought, although it may be overkill:

we can validate that the combination of the current and new MRC are in the right state

Reply:
I think I am in favor of this overkill. I'll make the update and we can evaluate if the complexity is worth it

Reply:
Updated. Only performing this check on update for now. I can add it on create as well but the mapping is slightly different